File: MercuryAPI-1.31.3.36/c/src/samples/licensekey.c
/**
 * Sample program that reads tags for a fixed period of time (500ms)
 * and prints the tags found.
 * @file LicenseKey.c
 */

#include <tm_reader.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdarg.h>
#include "serial_reader_imp.h"
#include <inttypes.h>

/* Enable this to use transportListener */
#ifndef USE_TRANSPORT_LISTENER
#define USE_TRANSPORT_LISTENER 0
#endif

#define usage() {errx(1, "Please provide valid reader URL, such as: reader-uri [--option <license operation>] [--key <licence key>]\n"\
                         "reader-uri : e.g., 'tmr:///COM1' or 'tmr:///dev/ttyS0/' or 'tmr://readerIP'\n"\
                         "<license operation> : a)set \n"\
                         "                      b)erase \n"\
                         "Example: tmr:///com4 --option set --key AB CD\n"\
                         "Example: tmr:///com4 --option erase\n");}

void errx(int exitval, const char *fmt, ...)
{
  va_list ap;

  va_start(ap, fmt);
  vfprintf(stderr, fmt, ap);

  exit(exitval);
}

void checkerr(TMR_Reader* rp, TMR_Status ret, int exitval, const char *msg)
{
  if (TMR_SUCCESS != ret)
  {
    errx(exitval, "Error %s: %s\n", msg, TMR_strerr(rp, ret));
  }
}

void serialPrinter(bool tx, uint32_t dataLen, const uint8_t data[],
                   uint32_t timeout, void *cookie)
{
  FILE *out = cookie;
  uint32_t i;

  fprintf(out, "%s", tx ? "Sending: " : "Received:");
  for (i = 0; i < dataLen; i++)
  {
    if (i > 0 && (i & 15) == 0)
    {
      fprintf(out, "\n         ");
    }
    fprintf(out, " %02x", data[i]);
  }
  fprintf(out, "\n");
}

void stringPrinter(bool tx,uint32_t dataLen, const uint8_t data[],uint32_t timeout, void *cookie)
{
  FILE *out = cookie;

  fprintf(out, "%s", tx ? "Sending: " : "Received:");
  fprintf(out, "%s\n", data);
}

void parseLicenseKey(uint8_t *license, uint8_t *length, char *args[])
{
  uint8_t tempLicense[100];
  uint8_t i, j, count;
  count = 0;
  for (i = 0; i < *length; i++)
  {
    for (j = 0; args[i][j] ;j++)
    {
      tempLicense[count++] = args[i][j];
    }
  }
  tempLicense[count] = '\0';
  count = 0;
  for (i = 0; tempLicense[i]; i = i + 2)
  {
    const char tempdata[2] = {tempLicense[i] , tempLicense[i+1]};
    sscanf(tempdata, "%hhx", &license[count++]);
  }
  license[count] = '\0';
  *length = count;
}

/**
 * Parse the license operation option 
 *
 * @param index : starting index of option
 * @param option : points to the variable where option need to store
 * @param args : pointer to command line argument array
*/
void parseLicenseOperationOption(int index, int *option, char *args[])
{
  /* parse license option provided by user */
  if((0x00 == strcmp("set", args[index + 1])) || (0x00 == strcmp("SET", args[index + 1])))
    *option = TMR_SET_LICENSE_KEY;
  else if((0x00 == strcmp("erase", args[index + 1])) || (0x00 == strcmp("ERASE", args[index + 1])))
    *option = TMR_ERASE_LICENSE_KEY;
  else
  {
    fprintf(stdout, "Unsupported license operation\n");
    usage();
  }    
}

/**
 * Calculates the license key length 
 *
 * @param index : starting index of license key
 * @param isOptionFound : option keyword found flag
 * @param argCount : arguments count
 * @param args : pointer to command line argument array
*/
int calculateLicenseKeyLength(int index, bool isOptionFound, int argCount, char *args[])
{ 
  int keyLen = 0;
  int nextIndex = 0;
  bool nextArgIsOption = 0;
  
  if(isOptionFound)
  {
      keyLen = argCount - 5;
  }
  else
  {
      nextIndex = index + 1;
    for(; (index < argCount) && !nextArgIsOption ; index++, nextIndex++)
    {
      if((0x00 == strcmp("--option", args[nextIndex])) || (0x00 == strcmp("--OPTION", args[nextIndex])))
        nextArgIsOption = true; 
      else if((nextIndex + 1) == argCount)
      {
        fprintf(stdout, "license operation option is not found\n");
        usage();
      }
      else
        keyLen++;
    }
    index--;
  }
  return keyLen;
}

int main(int argc, char *argv[])
{
  TMR_Reader r, *rp;
  TMR_Status ret;
  TMR_Region region;
  uint8_t *keyData = NULL;
  uint8_t buffer[100];
  uint8_t keyLength = 0;
  int i = 0;
  int keyIndex = 0;
  int option = 0;
  bool optionFound = false;
  bool keyFound = false;

  TMR_LicenseOperation licenseOperation;
#if USE_TRANSPORT_LISTENER
  TMR_TransportListenerBlock tb;
#endif

  if (argc < 4)
  {
    fprintf(stdout, "Not enough arguments.\n");
    usage();
  }
  else
  {
    for (i = 2; i < argc; i++)
    {
      /* check for license operation option */
      if((!optionFound) && ((0x00 == strcmp("--option", argv[i])) || (0x00 == strcmp("--OPTION", argv[i]))))
      {
        optionFound = true;
        parseLicenseOperationOption(i, &option, argv);
        i++; /* move index to next place */
      }

      /* check for license key */
      else if((!keyFound) && ((0x00 == strcmp("--key", argv[i])) || (0x00 == strcmp("--KEY", argv[i]))))
      {
        keyFound = true;
        keyIndex = i;  /* Store the key index */ 

        /* Calculate the license key length */
        keyLength = calculateLicenseKeyLength(i, optionFound, argc, argv);	
        i += keyLength;

        if(keyLength)
        {
          /* license key parsing */
          parseLicenseKey(buffer, &keyLength, argv + keyIndex + 1);
          keyData = buffer;
        }
        else
        { 
          fprintf(stdout, "license key not found\n");
          usage();
        }
      }
      else
      {
        fprintf(stdout, "Arguments are not recognized\n");
        usage();
      }
    }
    /* Error handling */
    if(!optionFound)
    { 
      fprintf(stdout, "license operation option is not found\n");
      usage();
    }
    else if ((option == 1) && (!keyFound))
    { 
      fprintf(stdout, "license key not found\n");
      usage();
    }
  }

  rp = &r;
  ret = TMR_create(rp, argv[1]);
  checkerr(rp, ret, 1, "creating reader");

#if USE_TRANSPORT_LISTENER

  if (TMR_READER_TYPE_SERIAL == rp->readerType)
  {
    tb.listener = serialPrinter;
  }
  else
  {
    tb.listener = stringPrinter;
  }
  tb.cookie = stdout;

  TMR_addTransportListener(rp, &tb);
#endif

  ret = TMR_connect(rp);
  checkerr(rp, ret, 1, "connecting reader");

  region = TMR_REGION_NONE;
  ret = TMR_paramGet(rp, TMR_PARAM_REGION_ID, &region);
  checkerr(rp, ret, 1, "getting region");

  if (TMR_REGION_NONE == region)
  {
    TMR_RegionList regions;
    TMR_Region _regionStore[32];
    regions.list = _regionStore;
    regions.max = sizeof(_regionStore)/sizeof(_regionStore[0]);
    regions.len = 0;

    ret = TMR_paramGet(rp, TMR_PARAM_REGION_SUPPORTEDREGIONS, &regions);
    checkerr(rp, ret, __LINE__, "getting supported regions");

    if (regions.len < 1)
    {
      checkerr(rp, TMR_ERROR_INVALID_REGION, __LINE__, "Reader doesn't supportany regions");
    }
    region = regions.list[0];
    ret = TMR_paramSet(rp, TMR_PARAM_REGION_ID, &region);
    checkerr(rp, ret, 1, "setting region");  
  }

  /* Manage license key */
  {
    TMR_uint8List key;
#if(1) /* Change to "if (0)" to provide key from code */
    key.list = keyData;
    key.len = key.max = keyLength;
#else
    uint8_t keyData1[] = {
      0xAB, 0xCD
    };
    key.list = keyData1;
    key.len = key.max = sizeof(keyData1) / sizeof(keyData1[0]);
#endif

    /* Set license operation */
    licenseOperation.option = option;
    licenseOperation.license = &key;

    printf("License Key operation started...\n");

    //ret = TMR_paramSet(rp, TMR_PARAM_LICENSE_KEY, &key);
    ret = TMR_paramSet(rp, TMR_PARAM_MANAGE_LICENSE_KEY, &licenseOperation);
    if (TMR_SUCCESS != ret)
    {
      fprintf(stderr, "Error license operation: %s\n", TMR_strerr(rp, ret));
    }
    else
    {
      printf("License operation succeeded.\n");
    }
  }

  // Report protocols enabled by current license key
  {
    TMR_TagProtocolList protocols;
    TMR_TagProtocol protocolData[16];
    int i;

    protocols.list = protocolData;
    protocols.max = sizeof(protocolData) / sizeof(protocolData[0]);
    protocols.len = 0;

    ret = TMR_paramGet(rp, TMR_PARAM_VERSION_SUPPORTEDPROTOCOLS, &protocols);
    checkerr(rp, ret, 1, "getting supported protocols");
    printf("Supported Protocols:\n");
    for (i=0; i<protocols.len; i++)
    {
      char* protocolName;

      switch (protocols.list[i])
      {
      case TMR_TAG_PROTOCOL_GEN2:
        protocolName = "GEN2";
        break;
      case TMR_TAG_PROTOCOL_ISO180006B:
        protocolName = "ISO18000-6B";
        break;
      case TMR_TAG_PROTOCOL_ISO180006B_UCODE:
        protocolName = "ISO18000-6B_UCODE";
        break;
      case TMR_TAG_PROTOCOL_IPX64:
        protocolName = "IPX64";
        break;
      case TMR_TAG_PROTOCOL_IPX256:
        protocolName = "IPX256";
        break;
      case TMR_TAG_PROTOCOL_ATA:
        protocolName = "ATA";
        break;
      default:
        protocolName = NULL;
        break;
      }

      if (NULL != protocolName)
      {
        printf("%s\n", protocolName);
      }
      else
      {
        printf("0x%02X\n", protocols.list[i]);
      }
    }
    printf("\n");
  }
 
  TMR_destroy(rp);
  return 0;
}

File: MercuryAPI-1.31.3.36/java/samples_nb/src/samples/LicenseKey.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package samples;
import com.thingmagic.*;
import com.thingmagic.Reader.LicenseOperation;

/**
 *
 * @author rsoni
 */
public class LicenseKey
{
    static SerialPrinter serialPrinter;
    static StringPrinter stringPrinter;
    static TransportListener currentListener;
    static LicenseOperation op = new LicenseOperation();
    static byte[] byteLic = new byte[]{};
    static String readerURI;
    static void usage()
   {
    System.out.printf("Usage: Please provide valid arguments, such as:\n"
                + "Licensekey [-v] [reader-uri] [--option] [set/erase] [--key] <license Key>\n" +
                  "-v  Verbose: Turn on transport listener\n" +
                  "reader-uri  Reader URI: e.g., \"tmr:///COM1\", \"tmr://astra-2100d3\"\n"
                + "--option: to select the options(1.set 2.erase)\n" +
                  "set: update given license e.g.,tmr:///COM4 --option set --key AB CD\n"+
                  "erase: erase: erase existing license e.g.,tmr:///COM4 --option erase '\n\n"
                + "e.g: tmr:///com1 --option set --key 112233 ; tmr://10.11.115.32 --option erase \n ");
    System.exit(1);
   }

   public static void setTrace(Reader r, String args[])
  {
      if (args[0].toLowerCase().equals("on"))
      {
          r.addTransportListener(Reader.simpleTransportListener);
          currentListener = Reader.simpleTransportListener;
      }
      else if (currentListener != null)
      {
          r.removeTransportListener(Reader.simpleTransportListener);
      }
  }

   static class SerialPrinter implements TransportListener
  {
    public void message(boolean tx, byte[] data, int timeout)
    {
      System.out.print(tx ? "Sending: " : "Received:");
      for (int i = 0; i < data.length; i++)
      {
        if (i > 0 && (i & 15) == 0)
        {
          System.out.printf("\n         ");
        }
        System.out.printf(" %02x", data[i]);
      }
      System.out.printf("\n");
    }
  }

  static class StringPrinter implements TransportListener
  {
    public void message(boolean tx, byte[] data, int timeout)
    {
      System.out.println((tx ? "Sending:\n" : "Receiving:\n") +
                         new String(data));
    }
  }
    public static void main(String argv[]) throws ReaderException
    {
        Reader r = null;
        int nextarg = 0;
        boolean trace = false;
        StringBuilder key = new StringBuilder(); 
        //String readerURI;
        int optionIndex = 0;
        int keyIndex = 0;
        int keyLength = 0;
        boolean optionFound = false;
        boolean keyFound = false;
        int startIndex = 0;
        int minArgc = 0;

        try
        {
            if (argv[nextarg].equals("-v"))
            {
                trace = true;
                nextarg++;
                startIndex = 2;
                minArgc = 4;
            }
            else
            {
                startIndex = 1;
                minArgc = 3;
            }

            readerURI = argv[nextarg];
            nextarg++;

            if (argv.length < minArgc)
            {
                usage();
            }
            else
            {
                for( nextarg = startIndex; nextarg < argv.length; nextarg++)
                {
                    /* check for license operation option */
                    if((!optionFound) && (argv[nextarg].equalsIgnoreCase("--option")))
                    {
                        optionFound = true;
                        /* parse license option provided by user */
                        parseLicenseOperationOption(nextarg, argv);
                        nextarg++;
                    }

                    /* check for license key */
                    else if((!keyFound) && (argv[nextarg].equalsIgnoreCase("--key")))
                    {
                        keyFound = true;
                        keyIndex = nextarg; /* Store the key index */ 

                        /* Calculate the license key length */
                        keyLength = calculateLicenseKeyLength(keyIndex, optionFound, argv.length, argv);
                        nextarg+= keyLength;

                        if(keyLength!=0)
                        {
                            /* parse the license key */
                            byteLic = parseLicenseKey(keyIndex, keyLength, argv);
                            op.key = byteLic;
                        }
                        else
                        {
                            System.out.println("License key not found");
                            usage();
                        }
                    }
                    else
                    {
                        System.out.println("Arguments are not recognized");
                        usage();
                    }
                }

                /* Error handling */
                if(!optionFound)
                { 
                  System.out.println("license operation option is not found");
                  usage();
                }
                else if ((op.option == Reader.LicenseOption.SET_LICENSE_KEY) && (!keyFound))
                { 
                  System.out.println("License key not found");
                  usage();
                }
            }
        }
        catch(Exception ex)
        {
            System.out.println("Exception: " + ex.getMessage());
        }

        try
        {
            r = Reader.create(readerURI);
            if (trace)
            {
                  setTrace(r, new String[]{"on"});
            }
            r.connect();
            if (Reader.Region.UNSPEC == (Reader.Region) r.paramGet("/reader/region/id")) {
                Reader.Region[] supportedRegions = (Reader.Region[]) r.paramGet(TMConstants.TMR_PARAM_REGION_SUPPORTEDREGIONS);
                if (supportedRegions.length < 1) {
                    throw new Exception("Reader doesn't support any regions");
                } else {
                    r.paramSet("/reader/region/id", supportedRegions[0]);
                }
            }

            //Uncomment this to only "Set" licensekey.      
            /*System.out.println("Set Protocol License Key started.\n");
            r.paramSet("/reader/licensekey", byteLic);
            System.out.println("Set Protocol License Key succeeded.\n");*/

            // Manage License key param supports both setting and erasing the license.
            System.out.println("License operation started.\n");
            r.paramSet(TMConstants.TMR_PARAM_MANAGE_LICENSE_KEY, op);
            System.out.println("License operation succeeded.\n");
            // Report protocols enabled by current license key
            TagProtocol[] protocolList = ( TagProtocol[])r.paramGet("/reader/version/supportedProtocols");
            System.out.println("Supported Protocols:" );
            for(TagProtocol p : protocolList)
            {
                System.out.println(p);
            }

        }
        catch (ReaderException re)
        {
            System.out.println("ReaderException: " + re.getMessage());
        }
        catch (Exception re)
        {
            System.out.println("Exception: " + re.getMessage());
        }
        finally
        {
            r.destroy();
        }
    }

    public static byte[] parseLicenseKey(int keyIndex, int keyLength, String[] argv)
    {
        /* parse license option provided by user */
        int keyStartPointer = keyIndex + 1;
        StringBuilder key = new StringBuilder();
        for (int i =0; i < keyLength; i++) {
            key.append(argv[keyStartPointer + i]);
        }
        String s = new String(key);
        byteLic = ReaderUtil.hexStringToByteArray(s);
        return byteLic;
    }
    
    public static void parseLicenseOperationOption(int index, String[] argv)
    {
        /* parse license option provided by user */
        String argument = argv[index + 1];
        if(argument.equalsIgnoreCase("SET"))
        {
            op.option = Reader.LicenseOption.SET_LICENSE_KEY;
        }
        else if(argument.equalsIgnoreCase("ERASE"))
        {
            op.option = Reader.LicenseOption.ERASE_LICENSE_KEY;
        }
        else
        {
            System.out.println("Unsupported license operation");
            usage();
        }
    }
    
    public static int calculateLicenseKeyLength(int index, boolean isOptionFound, int argCount, String[] argv )
    {
        int keyLen = 0;
        int nextIndex = 0;
        boolean nextArgIsOption = false;
        if(isOptionFound)
        {
            if(argv[0].equals("-v"))
            {
                keyLen = argCount - 5;
            }
            else
            {
                keyLen = argCount - 4;
            }
        }
        else
        {
            nextIndex = index + 1;
            for(; (index < argCount)&&(!nextArgIsOption) ; index++, nextIndex++)
            {
                if((argv[nextIndex].equalsIgnoreCase("--option")))
                {
                    nextArgIsOption = true;
                }
                else if((nextIndex + 1) == argCount)
                {
                    System.out.println("License operation option is not found");
                    usage();
                }
                else
                {
                    keyLen++;
                }
            }
            index--;
        }
        return keyLen;
    }

}
